sui_core/src/authority/authority_store.rs

```diff
@@ -632,6 +653,30 @@ impl<const ALL_OBJ_VER: bool> SuiDataStore<ALL_OBJ_VER> {
         self.batch_update_objects(write_batch, temporary_store, transaction_digest, None)
     }
 
+    pub fn update_gateway_state<S>(
```

Comment:
For transactions on owned objects, it should be so much easier to just re-execute. For shared objects though we might need to shared object input state to do that.